### PR TITLE
MULE-17010: OAuth uses wrong access token when multiple concurrent

### PR DESCRIPTION
**Pull request by elrodro83**
refresh requests are sent

### CODE REVIEW

File: src/test/java/org/mule/test/oauth/AbstractOAuthTestCase.java [hunk 1]
```diff
@@ -78,7 +88,27 @@ public void setupServices() throws Exception {
...
     when(httpEntity.getContent()).thenReturn(new ReaderInputStream(new StringReader("")));
     when(httpResponse.getEntity()).thenReturn(httpEntity);
-    when(httpClient.sendAsync(any(), any())).thenReturn(supplyAsync(() -> httpResponse));
+    when(httpClient.sendAsync(any(), any())).thenAnswer(invocation -> {
+
+      final CompletableFuture<HttpResponse> httpResponseFuture = new CompletableFuture<>();
+
+      httpClientCallbackExecutor.execute(() -> {
+        try {
+          sleep(10);
```
Comment (marianogonzalez) [inline src/test/java/org/mule/test/oauth/AbstractOAuthTestCase.java]: naaaaaaaaa

File: src/main/java/org/mule/service/oauth/internal/DefaultClientCredentialsOAuthDancer.java [hunk 2]
```diff
@@ -87,8 +88,8 @@ public DefaultClientCredentialsOAuthDancer(String clientId, String clientSecret,
   @Override
   public void start() throws MuleException {
     super.start();
-    // We use a reentrant instead of one from the lock factory because the local state of this object cannot be shared in the
-    // cluster.
+    // TODO MULE-17013 We use a reentrant instead of one from the lock factory because the local state of this object cannot be
```
Comment (marianogonzalez) [inline src/main/java/org/mule/service/oauth/internal/DefaultClientCredentialsOAuthDancer.java]: this was already done. We used to have a monitored object store in which we put an entry indicating that ta refresh token was in place. That entry had a relatively short TTL. That ObjectSTore was checked before dooing the refresh and if it was not empty, the refresh was not attempted and the current token was returned instead

File: src/test/java/org/mule/test/oauth/AbstractOAuthTestCase.java (hunk 1 above)
Comment (elrodro83) [reply src/test/java/org/mule/test/oauth/AbstractOAuthTestCase.java]: test code. It i were to do this with a latch, the countdown would need to be in production code :S

File: src/main/java/org/mule/service/oauth/internal/DefaultClientCredentialsOAuthDancer.java (hunk 2 above)
Comment (elrodro83) [reply src/main/java/org/mule/service/oauth/internal/DefaultClientCredentialsOAuthDancer.java]: that was ok for a blocking implementation, where you could just wait for the token to be put into the OS. For this non-blocking implementation, we need a notification to complete the returned `CompletableFuture`.

Comment (marianogonzalez) [reply src/main/java/org/mule/service/oauth/internal/DefaultClientCredentialsOAuthDancer.java]: what about a VM queue? Also, keep cloudhub limitations when defining this. We already have issues with the distributed locking. Need to come out with a solutiono that also works there

Comment (elrodro83) [reply src/main/java/org/mule/service/oauth/internal/DefaultClientCredentialsOAuthDancer.java]: a vm queue won't work with a cluster bigger than 2 nodes. Agree we need to further discuss the situation with CH.

**Pull request closed without merging**